machine/v1beta1/types_azureprovider.go

```diff
@@ -437,11 +481,38 @@ type DiskEncryptionSetParameters struct {
 // SecurityProfile specifies the Security profile settings for a
 // virtual machine or virtual machine scale set.
 type SecurityProfile struct {
-	// This field indicates whether Host Encryption should be enabled
-	// or disabled for a virtual machine or virtual machine scale
-	// set. Default is disabled.
+	// This field indicates whether Host Encryption should be enabled or disabled for a virtual
+	// machine or virtual machine scale set.
+	// This should be disabled when SecurityEncryptionType is set to DiskWithVMGuestState.
+	// Default is disabled.
 	// +optional
 	EncryptionAtHost *bool `json:"encryptionAtHost,omitempty"`
+	// SecurityType specifies the SecurityType of the virtual machine. It has to be set to any specified value to
+	// enable UEFISettings. The default behavior is: UEFISettings will not be enabled unless this property is set.
+	// +kubebuilder:validation:Enum=ConfidentialVM;TrustedLaunch
+	// +optional
+	SecurityType SecurityTypes `json:"securityType,omitempty"`
+	// UEFISettings specifies the security settings like secure boot and vTPM used while creating the virtual machine.
+	// +optional
+	UEFISettings *UEFISettings `json:"uefiSettings,omitempty"`
```

Comment:
This should be a discriminated union, that way you can encapsulate the relationship of x must only be set when y is set
```
vmSecurity:
  type: ConfidentialVM
  confidentialVM:
    uefi:
      ...
```

Reply:
Very good point, I thought about making the `SecurityProfile` struct a discriminated union, but there is also the `SecurityProfile.EncryptionAtHost` field. The latter can be set even when the `SecurityType` fielf is empty. Were you perhaps thinking about having an additional struct in the `SecurityProfile` struct, defined as a discriminated union?

Something like the following for example?

```go
type SecurityProfile struct {
  // encryptionAtHost indicates whether Host Encryption should be enabled or disabled for a virtual
  // machine or virtual machine scale set.
  // This should be disabled when SecurityEncryptionType is set to DiskWithVMGuestState.
  // Default is disabled.
  // +optional
  EncryptionAtHost *bool `json:"encryptionAtHost,omitempty"`
  
  // securitySettings ...
  Settings *SecuritySettings `json:"securitySettings,omitempty"`
}

// +union
type SecuritySettings struct {
  // securityType specifies the SecurityType of the virtual machine. It has to be set to any specified value to
  // enable UEFISettings. The default behavior is: UEFISettings will not be enabled unless this property is set.
  // +kubebuilder:validation:Enum=ConfidentialVM;TrustedLaunch
  // +unionDiscriminator
  // +required
  SecurityType SecurityTypes `json:"securityType,omitempty"`
  // uefiSettings specifies the security settings like secure boot and vTPM used while creating the virtual machine.
  // +optional
  UEFISettings *UEFISettings `json:"uefiSettings,omitempty"`
}
```

Reply:
Yes that's about right, but you need to include a separate third layer for confidential VM and TrustedLaunch, so it's 
```
securitySettings:
  securityType: ConfidentialVM
  confidentialVM:
    uefiSettings:
      ...
```
I'd probably look at how often we are duplicating words like security in this context, probably don't always need it for every field name

Reply:
As per our discussions [here](https://github.com/openshift/api/pull/1403#discussion_r1247626913), do you think we need to leave this bit closer to the Azure REST API (and the upstream implementation) or would you go for a change here? 

My view is the same as what we discussed before, i.e. we could risk future breaking changes.

Reply:
In this case, I think we can convert the API here to the API that upstream have implemented so I would go with the more structured API here, and then convert later

Reply:
Currently, the changes in this PR reflect the upstream implementation, i.e. [OS disk security profile](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go#L647-L665) and [VM security profile](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/types.go#L828-L863), adjusted to the OpenShift API conventions (e.g. bool pointers converted to string enums).

Reply:
Yeah that's ok, but I think the structure of the discriminated union I mentioned is preferable from the openshift conventions. Then later, when we come to convert MAPI resources to CAPI, the translation will just drop the union and go back to what upstream has

Reply:
@JoelSpeed is [this](https://github.com/openshift/api/pull/1403/files#diff-80356b3491fb4dddf4dd40c0fdbddff9391c70dbc3c3cb090740e629ca4098d5R503-R536) what you had in mind?

Reply:
Yes, that's the structure I had in mind, does this need another review now then?

Install config doesn't _have_ to be the same no, preferable though IMO

Reply:
> does this need another review now then?

Yes it does :)

> Install config doesn't have to be the same no, preferable though IMO

This makes sense, thanks for the info :+1: